### DISCUSSION TITLE
How to get an egui::Key enum variant from a string?

### DISCUSSION THREAD

**Discussion by Vince-LD** [emilk/egui#3368] [Q&A] [answered]
Hello,

I've made my first Rust program as well as my first native gui ever. I've added key shortcuts to perform actions in order to avoid mouse use as much as possible. The problem is that because my keyboard is in AZERTY layout, a QWERTY user would not be able to use the app comfortably. 
To try solve this issuen, I added a `config.ini` file in order to configure some app parameters as well as key mappings. However, I am enable to get the `egui::Key` enum variant from the string i get from the config file. I saw solutions with the strum `EnumString` and `serde_json` but none of those worked. My only solution is to shadow the `egui::Key` enum (with a macro?) to derive the `EnumString` or to code a function with a huge match statement in it to use against the string. Both  those options seem very sub-optimal but I have no idea how to do it otherwise. The only idea I have on how to do it simply is to directly add the `derive(EnumString)` in egui's source code.

Thanks for you help!

**Comment by YgorSouza** [comment, marked as answer]
You can use serde. egui derives its traits [here](https://github.com/emilk/egui/blob/33a0f50f6a80fe28a6430de7c70d6da79ebba2d5/crates/egui/src/data/input.rs#L677-L678). However, you have to enable the `serde` feature, and it seems that enabling it on eframe doesn't automatically enable it on egui. So you have to depend on egui explicitly to enable it, I guess. Or you can use the `persistence` feature from eframe, which enables serde on egui and a few other things, to allow it to save the full app state between sessions.

```toml
[dependencies]
eframe = { version = "0.22" }
egui = { version = "0.22", features = [ "serde"] }
serde_json = "1"
```

Then you can deserialize the key from a json string, for example.

```rs
use std::collections::BTreeMap;

fn main() -> Result<(), eframe::Error> {
    eframe::run_native(
        "My egui App",
        Default::default(),
        Box::new(|_cc| Box::<MyApp>::default()),
    )
}

#[derive(Default)]
struct MyApp {}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            let json = serde_json::to_string(&egui::Key::A).unwrap();
            let key: egui::Key = serde_json::from_str(&json).unwrap();
            ui.label(format!("JSON: {json}"));
            ui.label(format!("Enum: {key:?}"));
            let map: BTreeMap<egui::Key, egui::Key> =
                [(egui::Key::A, egui::Key::Q), (egui::Key::Z, egui::Key::W)].into();
            let json = serde_json::to_string(&map).unwrap();
            let map: BTreeMap<egui::Key, egui::Key> = serde_json::from_str(&json).unwrap();
            ui.label(format!("JSON: {json}"));
            ui.label(format!("Map {map:?}"));
        });
    }
}
```

![image](https://github.com/emilk/egui/assets/43298013/85b018c2-9528-4b8b-b7e4-12b53c458d25)

I don't know how serde plays with INI files, but you might want to use JSON or RON, which are far more common. You can create a type containing all the settings for your app (including things like the map from the example above), and load it from a JSON or RON file using serde. That's what eframe itself does when you enable the `persistence` feature. But if you really want to use INI and deserialize only the Key as JSON, make sure it is in quotes like `"A"`, so it's a valid JSON string.

**Reply by Vince-LD** [reply]
Hi!
Thank you for that greatly comprehensive answer! 

I had no idea there was a feature to enable... In that kind of situation, what should I look for? Is it just something I would have to know?

To answer your  point on the `ini` format: I have other simple parameters I would like to be customizable without having to recompile the code and the parser I am using from the ini crate is great for my needs. My users won't be very technical (maybe not at all) so a json defining multiple structs may be a bit too overkill haha

I'll make another post when I am able to implement your solution.

**Reply by Vince-LD** [reply]
It worked perfectly! Thanks again 😄

**Reply by YgorSouza** [reply]
Crates usually generate their docs with all features enabled, so you can look through the [docs](https://docs.rs/egui/latest/egui/enum.Key.html#trait-implementations) to see what is available, and then in most cases you have to click on the source link to see if it is gated behind a feature. It is possible to add markers in the docs indicating which features have to be enabled for each given item. You can see that in the [tokio](https://docs.rs/tokio/latest/tokio/#modules) docs for example. But AFAIK this isn't a stable feature of the docs and not many crates use it.

For serde in particular, almost every crate that declares types implements `Serialize` and `Deserialize`, and almost every crate puts that behind a feature called `serde`. It is actually a recommendation from the [Rust API guidelines](https://rust-lang.github.io/api-guidelines/interoperability.html#c-serde), so you can basically count on it for every crate you come across, and if it isn't there, you can open an issue about it.
